### ISSUE TITLE
Flexible input shape for map method in class RandomFourierFeatureMapper

### ISSUE THREAD

**Issue by dchatterjee172** [tensorflow/tensorflow#15771]
This is a feature request. In tensorflow r1.4, Class RandomFourierFeatureMapper, map method,
The shape of the input must be, [batch_size, self._output_dim].   

There can be scenarios where in each batch there are multiple training point, essentially i'm proposing a scenario where the input shape  must be  [batch_size, x ,self._output_dim].   

Which is not possible in current API.        
In the matmul we can see that it is mentioned, "The inputs must, following any transpositions, be tensors of rank >= 2 where the inner 2 dimensions specify valid matrix multiplication arguments, and any further outer dimensions match." which is in-fact very flexible.

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce

**Comment by dchatterjee172** [comment]
@tensorflowbutler As this is a feature request, not a bug, I think those fields are irrelevant here.

**Comment by tensorflowbutler** [comment]
Nagging Awaiting Response: It has been 14 days with no activityand the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
The original poster has replied to this issue after the stat:awaiting response label was applied.

**Comment by petrosmol** [comment]
@dchatterjee172 
Thanks for submitting this feature request. Just to clarify:
The current operation does support multiple training points per batch but each data point is a 1-dim tensor. However, I could definitely see a scenario where each data point is a multi-dim tensor. For instance batch of 2-D images, in which case input tensor would be 3-D ([batch_size, i, j]). Is this the use case you have in mind?

**Comment by dchatterjee172** [comment]
@petrosmol 
Yes that's a possible case. But there can be many possible other use cases. For example in machine translation or machine comprehension you can have, [batch_size, words ,self._output_dim]. In my opinion as kernel tricks are essentially important part of machine learning, [batch_size, x ,self._output_dim] this input format will generalize this kernel approximation method.

**Comment by petrosmol** [comment]
Just to clarify, you are referring to self._input_dim not self._output_dim in this thread right? For instance you want to allow shape [batch_size, x, self._input_dim] for ]the input dimension, correct?

In any case, I can add a TODO, but I don't have an ETA for this.

**Comment by dchatterjee172** [comment]
@petrosmol 
Ya you are right [batch_size, x, self._input_dim]. 
Ya that's fine

**Comment by dchatterjee172** [comment]
If you can add a "contributions welcome" label then that would be great

**Comment by dchatterjee172** [comment]
@petrosmol 
The output should be also [batch_size, x, self._output_dim]. For example if you have a sentence, where x is the sentence length or how many words the sentence has, and self._output_dim will be the size of word vectors.
